core/src/test/java/org/everit/json/schema/ConditionalSchemaTest.java

```diff
@@ -0,0 +1,193 @@
+package org.everit.json.schema;
+
+import org.junit.Test;
+
+import static org.everit.json.schema.TestSupport.buildWithLocation;
+
+public class ConditionalSchemaTest {
+
+    private static final StringSchema MAX_LENGTH_STRING_SCHEMA = StringSchema.builder().maxLength(4).build();
+    private static final StringSchema MIN_LENGTH_STRING_SCHEMA = StringSchema.builder().minLength(6).build();
+    private static final StringSchema PATTERN_STRING_SCHEMA = StringSchema.builder().pattern("f.*o").build();
+
+    // only if
+
+    @Test
+    public void onlyIfSuccessEvenIfDataIsInvalidAgainstSubschema() {
+        ConditionalSchema.builder().ifSchema(StringSchema.builder().maxLength(2).build()).build().validate("foo");
+    }
+
+    @Test
+    public void onlyIfSuccess() {
+        ConditionalSchema.builder().ifSchema(MAX_LENGTH_STRING_SCHEMA).build().validate("foo");
+    }
+
+    // only then
+
+    @Test
+    public void onlyThenSuccessEvenIfDataIsInvalidAgainstSubschema() {
+        ConditionalSchema.builder().thenSchema(StringSchema.builder().maxLength(2).build()).build().validate("foo");
+    }
+
+    @Test
+    public void onlyThenSuccess() {
+        ConditionalSchema.builder().thenSchema(MAX_LENGTH_STRING_SCHEMA).build().validate("foo");
+    }
+
+    // only else
+
+    @Test
+    public void onlyElseSuccessEvenIfDataIsInvalidAgainstSubschema() {
+        ConditionalSchema.builder().elseSchema(StringSchema.builder().maxLength(1).build()).build().validate("foo");
+    }
+
+    @Test
+    public void onlyElseSuccess() {
+        ConditionalSchema.builder().elseSchema(MAX_LENGTH_STRING_SCHEMA).build().validate("foo");
+    }
+
+    // if-then
+
+    @Test
+    public void ifSubschemaSuccessThenSubschemaFailure() {
```

Comment:
If you pass a `Schema.Builder` to the `failureOf()` then it will call both `buildWithLocation()` and also set an expected pointer. You also don't have to explicitly pass the `expectedViolatedSchema` if it is the same as the root schema (it matters only when we check sub-schema failures). So this test can be written as 

```
ConditionalSchema.Builder subject = ConditionalSchema.builder().ifSchema(MAX_LENGTH_STRING_SCHEMA).thenSchema(PATTERN_STRING_SCHEMA);
        TestSupport.failureOf(subject)
                .expectedKeyword("then")
                .input("bar")
                .expect();
```

But if you prefer being more explicit then you can leave it as-is.